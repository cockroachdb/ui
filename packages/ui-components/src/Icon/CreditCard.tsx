import React, { useMemo, SVGProps } from "react";
import classnames from "classnames/bind";
import get from "lodash/get";

import { Cards } from "@cockroachlabs/icons";
import styles from "./card.module.scss";

import objectToClassNames from "../utils/objectToClassnames";

export type CreditCardSize = "tiny" | "small" | "medium" | "large";

type OwnCreditCardProps = {
  creditCardName?: keyof typeof Cards;
  size?: CreditCardSize;
};

type CreditCardProps = SVGProps<SVGElement> & OwnCreditCardProps;

const cx = classnames.bind(styles);

export const CreditCard = ({
  creditCardName = "CreditCard",
  size = "small",
  className,
  ...props
}: CreditCardProps) => {
  const classNames = useMemo(
    () => cx("card", objectToClassNames({ size }), className),
    [className, size],
  );
  const Element = get(Cards, creditCardName, null);

  if (Element === null) {
    return null;
  }

  return <Element className={classNames} {...props} />;
};
