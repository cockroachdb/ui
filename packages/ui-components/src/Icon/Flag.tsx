import React, { useMemo, SVGProps } from "react";
import classnames from "classnames/bind";
import get from "lodash/get";

import { Flags } from "@cockroachlabs/icons";
import styles from "./card.module.scss";

import objectToClassNames from "../utils/objectToClassnames";

export type FlagSize = "tiny" | "small" | "medium" | "large";

type OwnBaseFlagProps = {
  size?: FlagSize;
};

type OwnFlagPropsWithFlagName = {
  flagName: keyof typeof Flags;
  countryCode?: string;
};

type OwnFlagPropsWithCountryCode = {
  flagName?: keyof typeof Flags;
  countryCode: string;
};

type FlagProps = SVGProps<SVGElement> &
  OwnBaseFlagProps &
  (OwnFlagPropsWithFlagName | OwnFlagPropsWithCountryCode);

const cx = classnames.bind(styles);

export const flagNameCountryCodeMap: Array<{
  code: string;
  flag: keyof typeof Flags;
}> = [
  { code: "aus", flag: "Australia" },
  { code: "bel", flag: "Belgium" },
  { code: "bhr", flag: "Bahrain" },
  { code: "bra", flag: "Brazil" },
  { code: "can", flag: "Canada" },
  { code: "che", flag: "Switzerland" },
  { code: "deu", flag: "Germany" },
  { code: "fin", flag: "Finland" },
  { code: "fra", flag: "France" },
  { code: "gbr", flag: "UnitedKingdom" },
  { code: "hkg", flag: "HongKong" },
  { code: "ind", flag: "India" },
  { code: "idn", flag: "Indonesia" },
  { code: "irl", flag: "Ireland" },
  { code: "jpn", flag: "Japan" },
  { code: "kor", flag: "Korea" },
  { code: "nld", flag: "Netherlands" },
  { code: "pol", flag: "Poland" },
  { code: "sgp", flag: "Singapore" },
  { code: "swe", flag: "Sweden" },
  { code: "twn", flag: "Taiwan" },
  { code: "usa", flag: "Usa" },
  { code: "zaf", flag: "SouthAfrica" },
];

const getFlagNameFromCountryCode = (countryCode: string): keyof typeof Flags =>
  flagNameCountryCodeMap.find((m) => m.code === countryCode).flag;

export const Flag = ({
  flagName: flagNameProp,
  countryCode,
  size = "small",
  className,
  ...props
}: FlagProps) => {
  const classNames = useMemo(
    () => cx("card", objectToClassNames({ size }), className),
    [className, size],
  );
  const flagName = flagNameProp || getFlagNameFromCountryCode(countryCode);
  const Element = get(Flags, flagName, null);

  if (Element === null) {
    return null;
  }

  return <Element className={classNames} {...props} />;
};
