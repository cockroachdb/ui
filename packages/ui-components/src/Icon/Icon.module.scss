.icon {
  display: inline-flex;
  margin: auto;

  &:disabled {
    pointer-events: none;
  }
}

.size-large {
  height: 64px;
  width: 64px;
}

.size-medium {
  height: 24px;
  width: 24px;
}

.size-default {
  height: 16px;
  width: 16px;
}

.size-small {
  height: 12px;
  width: 12px;
}

.size-tiny {
  height: 8px;
  width: 8px;
}

.tint-blue {
  fill: #0788ff;
}

.tint-green {
  fill: #37a806;
}

.tint-purple {
  fill: #542abd;
}

.tint-red {
  fill: #ff3b4e;
}

.tint-white {
  fill: white;
}

.tint-neutral {
  fill: #242a35;
}

.tint-inherit {
  fill: currentColor;
}
