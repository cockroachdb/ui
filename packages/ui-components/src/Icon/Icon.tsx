import React, { FunctionComponent } from "react";
import classNames from "classnames/bind";
import get from "lodash/get";

import * as Icons from "@cockroachlabs/icons";
import styles from "./Icon.module.scss";

type IconSize = "large" | "medium" | "default" | "small" | "tiny";
type IconTint =
  | "blue"
  | "green"
  | "purple"
  | "red"
  | "white"
  | "neutral"
  | "inherit";

export interface IconProps {
  iconName: keyof typeof Icons;
  size?: IconSize;
  tint?: IconTint;
}

const cx = classNames.bind(styles);

const Icon: FunctionComponent<IconProps> = ({
  iconName,
  size = "default",
  tint = "neutral",
  ...unhandledProps
}) => {
  const classnames = cx("icon", {
    "size-large": size === "large",
    "size-medium": size === "medium",
    "size-default": size === "default",
    "size-small": size === "small",
    "size-tiny": size === "tiny",

    "tint-blue": tint === "blue",
    "tint-green": tint === "green",
    "tint-purple": tint === "purple",
    "tint-red": tint === "red",
    "tint-white": tint === "white",
    "tint-neutral": tint === "neutral",
    "tint-inherit": tint === "inherit",
  });
  const Element = get(Icons, iconName, null);

  if (Element === null) {
    return null;
  }

  return <Element className={classnames} {...unhandledProps} />;
};

export default Icon;
