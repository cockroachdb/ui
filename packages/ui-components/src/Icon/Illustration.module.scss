.illustration {
  max-height: 152px;
  max-width: 240px;
  width: 100%;
}
