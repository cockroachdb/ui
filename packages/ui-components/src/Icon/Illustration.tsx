import React, { useMemo, SVGProps } from "react";
import classnames from "classnames/bind";

import { Illustrations } from "@cockroachlabs/icons";
import styles from "./Illustration.module.scss";

type OwnIllustrationProps = {
  illustrationName: keyof typeof Illustrations;
};

export type IllustrationProps = SVGProps<SVGSVGElement> & OwnIllustrationProps;

const cx = classnames.bind(styles);

export const Illustration = ({
  illustrationName,
  className,
  ...props
}: IllustrationProps) => {
  const classNames = useMemo(() => cx("illustration", className), [className]);

  const Element = Illustrations[illustrationName];

  if (Element == null) {
    return null;
  }

  return <Element className={classNames} {...props} />;
};
