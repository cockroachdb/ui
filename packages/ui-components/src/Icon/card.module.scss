.size-tiny {
  height: 16px;
  width: 16px;
}
.size-small {
  height: 24px;
  width: 24px;
}
.size-medium {
  height: 28px;
  width: 28px;
}
.size-large {
  height: 32px;
  width: 32px;
}
