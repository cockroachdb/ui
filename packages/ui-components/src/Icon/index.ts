export * from "./Icon";
export * from "./Pictogram";
