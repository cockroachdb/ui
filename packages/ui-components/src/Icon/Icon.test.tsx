import React from "react";
import { shallow } from "enzyme";
import { MinusCircle } from "@cockroachlabs/icons";

import Icon, { IconProps } from "./Icon";

const defaultProps = {
  iconName: "Plus",
};

const renderSubject = (overrideProps?: IconProps) => {
  const props = Object.assign(defaultProps, overrideProps);
  return shallow(<Icon {...props} />);
};

test("Icon should render given an iconName", () => {
  const wrapper = renderSubject();
  expect(wrapper).toMatchSnapshot();
});

test("Icon name should correspond to an SVG", () => {
  const wrapper = renderSubject({ iconName: "MinusCircle" });
  expect(wrapper.find(MinusCircle).length).toBe(1);
});

describe("Icon size prop", () => {
  test("will be defaulted", () => {
    const wrapper = renderSubject();
    expect(wrapper.prop("className")).toContain("size-default");
  });

  test("will change the size of the Icon", () => {
    const wrapper = renderSubject();

    wrapper.setProps({ size: "large" });
    expect(wrapper.prop("className")).toContain("size-large");

    wrapper.setProps({ size: "medium" });
    expect(wrapper.prop("className")).toContain("size-medium");

    wrapper.setProps({ size: "small" });
    expect(wrapper.prop("className")).toContain("size-small");

    wrapper.setProps({ size: "tiny" });
    expect(wrapper.prop("className")).toContain("size-tiny");
  });
});

describe("Icon tint prop", () => {
  test("defaults to neutral", () => {
    const wrapper = renderSubject();
    expect(wrapper.prop("className")).toContain("tint-neutral");
  });

  test("changes the tint of an icon", () => {
    const wrapper = renderSubject();

    wrapper.setProps({ tint: "blue" });
    expect(wrapper.prop("className")).toContain("tint-blue");

    wrapper.setProps({ tint: "green" });
    expect(wrapper.prop("className")).toContain("tint-green");

    wrapper.setProps({ tint: "purple" });
    expect(wrapper.prop("className")).toContain("tint-purple");

    wrapper.setProps({ tint: "red" });
    expect(wrapper.prop("className")).toContain("tint-red");

    wrapper.setProps({ tint: "white" });
    expect(wrapper.prop("className")).toContain("tint-white");

    wrapper.setProps({ tint: "neutral" });
    expect(wrapper.prop("className")).toContain("tint-neutral");

    wrapper.setProps({ tint: "inherit" });
    expect(wrapper.prop("className")).toContain("tint-inherit");
  });
});
