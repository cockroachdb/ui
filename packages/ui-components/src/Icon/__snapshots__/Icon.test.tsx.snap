// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render given an iconName 1`] = `
<Plus
  className="icon size-default tint-neutral"
/>
`;
